File: Combination Sum III/combination-sum-iii.py
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        def handle(index, n, k, current_sum, temp, res):
            if current_sum == n and k == 0:
                res.append(temp[:])
                return
            if current_sum > n or k == 0:
                return
            
            for i in range(index, 10):
                temp.append(i)
                handle(i + 1, n, k - 1, current_sum + i, temp, res)
                temp.pop()
        
        res = []
        handle(1, n, k, 0, [], res)
        return res

File: Combination Sum III/combination-sum-iii.cpp
class Solution {
public:
    void handle(int index, int n, int& k, int& sum, vector<int>& temp,
                vector<vector<int>>& res) {
        if (sum == n && k == 0) {
            res.push_back(temp);
            return;
        }
        if (sum > n || k == 0)
            return;
        for (int i = index; i <= 9; i++) {
            sum += i;
            k--;
            temp.push_back(i);
            handle(i + 1, n, k, sum, temp, res);
            temp.pop_back();
            sum-=i;
            k++;
        }
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<int> temp;
        vector<vector<int>> res;
        int sum = 0;
        handle(1, n, k, sum, temp, res);
        return res;
    }
};